include/deal.II/lac/petsc_snes.h

```diff
@@ -0,0 +1,477 @@
+   *    ...
+   *    explicit MatrixType(Mat);
+   *    ...
+   *    Mat & petsc_matrix();
+   *    ...
+   * @endcode
+   *
+   * In particular, the supported types are the ones that can *wrap*
+   * PETSc's native vector and matrix classes, that are able to modify
+   * them in place, and that can return PETSc native types when requested.
+   *
+   * To use the solvers the user needs to provide the implementation of $F$ via
+   * the NonlinearSolver::residual callback.
+   *
+   * The default linearization procedure of a solver instantiated with
+   * this class consists in using Jacobian-Free-Newton-Krylov; the action of
+   * tangent matrices inside a linear solver process are approximated via
+   * matrix-free finite-differencing of the nonlinear residual equations.
+   * For details, consult the [PETSc
+   * manual](https://petsc.org/release/manual/snes/#sec-nlmatrixfree).
+   *
+   * In alternative, users can also provide the implementation of the
+   * Jacobian. This can be accomplished in two ways:
+   *  - a-la-PETSc style using NonlinearSolver::jacobian
+   *  - a-la-Deal.II style using NonlinearSolver::setup_jacobian and
+   * NonlinearSolver::solve_for_jacobian_system
+   *
+   * In case both approaches are coded, the deal.II style
+   * will be used.
+   *
+   * The second approach is more in style with the deal.II philosophy
+   * and it also allows command line customization, like for example,
+   * @code
+   * ./myApp -snes_type newtontr -ksp_type cg
+   * @endcode
+   * in case the user wants to change the default nonlinear solver to
+   * a trust region solver and iterate on the tangent system with CG,
+   * still using NonlinearSolver::solve_for_jacobian_system as a preconditioner.
+   *
+   * The first approach has instead the advantage that only the matrix assembly
+   * procedure has to be coded, thus allowing quicker implementations and
+   * faster turnaround for experimenting with linear solver preconditioning
+   * configurations via command line customizations, like for example,
+   * @code
+   * ./myApp -ksp_type cg -pc_type gamg
+   * @endcode
+   *
+   * In case the nonlinear equations are derived from energy minimization
+   * arguments, it may be beneficial to perform linesearch or test trust-region
+   * model reductions using the energy functional. In such cases, users can
+   * set an optional NonlinearSolver::energy callback.
+   */
+  template <typename VectorType  = PETScWrappers::VectorBase,
+            typename PMatrixType = PETScWrappers::MatrixBase,
+            typename AMatrixType = PMatrixType>
+  class NonlinearSolver
+  {
+  public:
+    /**
+     * Type that holds real-valued numbers.
+     *
+     * Used to represent norms.
+     */
+    using real_type = PetscReal;
+
+    /**
+     * Constructor.
+     */
+    NonlinearSolver(const NonlinearSolverData &data     = NonlinearSolverData(),
+                    const MPI_Comm &           mpi_comm = PETSC_COMM_WORLD);
```

Comment:
Normally we don't default the communicator. What is the reason for that? Is it always used or just under certain circumstances?

Reply:
The constructor creates the `SNES` object that needs the communicator.

Reply:
I think these two things are orthogonal.

In the issue you linked, explicitly using MPI_COMM_WORLD in code is wrong, period.

Using it as a default argument is a design choice. For example, what we do in `petsc4py` for default arguments is 
to have in all API calls `comm=None` (which may translate to `MPI_COMM_NULL` in the C++ world).

We then handle default choices with https://petsc.org/main/petsc4py/reference/petsc4py.PETSc.Sys.html#petsc4py.PETSc.Sys.getDefaultComm, which , by default, returns the world communicator unless the user specified differently (and this can be dynamic) with https://petsc.org/main/petsc4py/reference/petsc4py.PETSc.Sys.html#petsc4py.PETSc.Sys.setDefaultComm.